web/src/settings_ui.ts

```diff
@@ -19,33 +19,42 @@ export const strings = {
     saving: $t({defaultMessage: "Saving"}),
 };
 
+type DoSettingsChangeOptions = {
+    failure_msg_html?: string;
+    sticky?: boolean;
+    success_msg_html?: string;
+    $error_msg_element?: JQuery<HTMLElement>;
+    error_continuation?(xhr?: JQuery.jqXHR): void;
+    success_continuation?(response_data: JQuery.AjaxSettings["data"]): void;
+};
+
 // Generic function for informing users about changes to the settings
 // UI.  Intended to replace the old system that was built around
 // direct calls to `ui_report`.
 export function do_settings_change(
-    request_method,
-    url,
-    data,
-    status_element,
+    request_method: (options: JQuery.AjaxSettings) => JQuery.jqXHR | undefined,
+    url: string,
+    data: JQuery.AjaxSettings["data"],
+    status_element: JQuery<HTMLElement>,
     {
-        success_msg_html = strings.success_html,
-        failure_msg_html = strings.failure_html,
-        success_continuation,
         error_continuation,
+        failure_msg_html = strings.failure_html,
         sticky = false,
+        success_continuation,
+        success_msg_html = strings.success_html,
         $error_msg_element,
-    } = {},
-) {
+    }: DoSettingsChangeOptions = {},
+): void {
     const $spinner = $(status_element).expectOne();
     $spinner.fadeTo(0, 1);
     loading.make_indicator($spinner, {text: strings.saving});
     const remove_after = sticky ? undefined : 1000;
     const appear_after = 500;
 
-    request_method({
+    void request_method({
         url,
         data,
-        success(response_data) {
+        success(response_data: JQuery.AjaxSettings["data"]) {
```

Comment:
`JQuery.AjaxSettings["data"]` is not a helpful type, it just resolves to essentially `Record<string, any>`. We should declare this as `unknown` and use [Zod](https://zod.dev/) to safely narrow the type to something useful at runtime.

Reply:
@andersk `JQuery.AjaxSettings["data"]` represents the generic shape of data that you'd send anywhere when making an ajax request. So, it will never result in an error. If we declare this as `unknown` wouldn't other modules that are using this module complain when they're migrated to typescript? `unknown` is the antithesis of `any`, with unknown we wouldn't be able to pass in any value at all.

I had considered the fact that this function is accepting a generic object which can literally be anything, but right now the shape of data is not clearly defined even though this module seems related to `settings` functionality. I tried to narrow down the types but the other modules which depend on this one also have `any` as the type for everything, so it really wasn't possible. I had mentioned this in the [commit message](https://github.com/zulip/zulip/pull/24813/commits/65bdd66ff4693e68d0136fc0c918856ed4e3936d). As we gradually convert other modules dependent on this one, we'd start getting information on more specific types, and this module can then similarily be made to accept stricter, pre-defined shapes of values.

Zulip doesn't currently use `zod` for validation/transformation and type generation, so idk what you meant by that.

Reply:
We certainly shouldn't block this on integrating `Zod`, but I think the claim is marking this as `unknown` is better than a useless type?

I'm not sure it matters much; when we integrate Zod, we're definitely going to end up looking at this code paths, and we're surely going to spend time adjusting this function when doing so.

Reply:
@timabbott I thought using `unknown` wouldn't let us pass any argument but that's not the case. The `unknown` thing is the same as `any` for passing arguments, it only matters when the argument is used inside function body. So, inside the body it'll be `unknown` and we'll have to do strict type checking to work on it.

Currently it's just being passed to `success_continuation` function. That function will need to be provided as argument when `do_settings_change` is used. So, inside the definition of `success_continuation` function, we'll have to work with `response_data` of `unknown` type.

Reply:
@andersk @timabbott Made the `response_type` `unknown` and added a more specific type for `options` object for `request_method` argument.